labgrid/driver/usbvideodriver.py

```diff
@@ -38,23 +44,36 @@ def select_caps(self, hint=None):
         raise InvalidConfigError("Unkown video format {} for device {:04x}:{:04x}".format(
             variant, self.video.vendor_id, self.video.model_id))
 
-    def get_pipeline(self):
+    def get_pipeline(self, path, caps, controls=None):
         match = (self.video.vendor_id, self.video.model_id)
         if match == (0x046d, 0x082d):
-            return "v4l2src device={} ! {} ! h264parse ! fdsink"
-        if match == (0x046d, 0x0892):
-            return "v4l2src device={} ! {} ! decodebin ! vaapipostproc ! vaapih264enc ! h264parse ! fdsink"  # pylint: disable=line-too-long
-        raise InvalidConfigError("Unkown USB video device {:04x}:{:04x}".format(*match))
+            controls = controls or "focus_auto=1"
+            inner = "h264parse"
+        elif match == (0x046d, 0x0892):
+            inner = "decodebin ! vaapipostproc ! vaapih264enc ! h264parse"
+        elif match == (0x534d, 0x2109):
+            inner = None  # just forward the jpeg frames
+        else:
+            raise InvalidConfigError("Unkown USB video device {:04x}:{:04x}".format(*match))
```

Comment:
You could allow arbitrary cameras with something like this:
`v4l2src device=/dev/video0 ! 'image/jpeg;video/x-h264' ! parsebin ! matroskamux ! streamable=true ! fdsink`

Note that the order of the caps define the priority. So in this example, jpeg will be used if the camera supports both jpeg and h264.

Reply:
And you need some caps here, otherwise some raw format will most likely be chosen.

Reply:
We currently select the caps explicitly from low/mid/high settings.

Reply:
You can use ranges there: `image/jpeg,width=[1,1280]height=[1,720],framerate=[1/1,15/1]`
From what I can tell, it tries to use the larges values that fit into the range.

Reply:
OK. But for now, I'd prefer to use tested and explicit values there.